pypuf/experiments/experimenter.py

```diff
@@ -0,0 +1,95 @@
+import multiprocessing
+import logging
+from pypuf.tools import setup_logger
+
+# this semaphore should be only used by experiment.base.Experiment
+_semaphore = None
+# this queue should be only used by experiment.base.Experiment in oder to gather experiment output
+_queue = None
+
+
+def output_queue():
+    """
+        This function should be used to get the thread safe queue singleton.
+        It should protect from mess up the parallel execution.
+    :return: queue.Queue
+    """
+    global _queue
+    if _queue is None:
+        _queue = multiprocessing.Queue()
+    return _queue
+
+
+def semaphore():
+    """
+        This function should be used to get the semaphore to coordinate the parallel execution.
+        It should protect from mess up the parallel execution.
+    :return: multiprocessing.BoundedSemaphore
+    """
+    global _semaphore
+    _cpu_count = multiprocessing.cpu_count()
+    if _semaphore is None:
+        _semaphore = multiprocessing.BoundedSemaphore(_cpu_count)
+    return _semaphore
+
+
+class Experimenter(object):
+    # This static variable holds the message which signals the output thread that all processes are finished.
+    FINISHED = 'EXPERIMENTER FINISHED'
+
+    def __init__(self, log_name, experiments):
+        """
+        :param experiments: A list of pypuf.experiments.experiment.base.Experiment
+        :param log_name: A unique file path where to output should be logged.
+        """
+        self.experiments = experiments
+        self.log_name = log_name
+        setup_logger(self.log_name, self.log_name)
+        self.logger = logging.getLogger(self.log_name)
+        self.out_queue = output_queue()
+
+    def gather_output(self, out_queue):
+        """
+            This method is used to gather all the outputs from experiments.
+        """
+        msg = out_queue.get()
+        # log the experiment output
+        while msg != Experimenter.FINISHED or out_queue.qsize() > 0:
+            # log the output
+            self.logger.info(msg)
+            msg = out_queue.get()
+
+    def start_jobs(self, out_queue):
+        """
+            This method is used a master process with starts worker in order to execute experiments.
+        :param out_queue: The queue to share outputs from experiments.
+        """
+        sema = semaphore()
+
+        jobs = []
+
+        # start experiment jobs
+        for exp in self.experiments:
+            job = multiprocessing.Process(target=exp.execute, args=(out_queue,))
+            # is required to cap the amount of running experiments
+            sema.acquire()
+            job.start()
+            jobs.append(job)
+
+        # wait for all processes to be finished
+        for job in jobs:
+            job.join()
+
+        # signal the output thread to be done
+        out_queue.put(Experimenter.FINISHED)
+
+    def run(self):
+        # master process
+        master_process = multiprocessing.Process(target=self.start_jobs, args=(self.out_queue,))
```

Comment:
Why do we need `run()` and `start_jobs()` in different processes? Can we not just call `start_jobs()` directly?